### PR TITLE
Markov

### PR DESCRIPTION
**Pull request by arcadus**
Added markov chat bot functionality
Cleaned up the read_channel_msgs function in pybotX.py a bit
Rewrote the markov function in pybotX.py

@dezka @n0arch

### CODE REVIEW

**Comment by n0arch** [comment]
can you squash this?

**Comment by arcadus** [comment]
squash'd

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,3 @@
+tw.txt
+markov.pyc
+pybot.json
```
Comment (n0arch) [inline .gitignore]: you don't need to commit this file

File: markov.py [hunk 2]
```diff
@@ -0,0 +1,186 @@
+#!/usr/bin/env python
+'''
+Markov chain class for generating sentences for PyBot
+
+read/write methods for reading and writing the brain from/to a json file.
+
+string2list method is to ensure the markov object is using a list of
```
Comment (n0arch) [inline markov.py]: break these method descriptions out of here and put them in the function itself so you don't have 2 scroll up to find a description

File: markov.py [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+        self.next_word() to provide the seed word. Would be nice to somehow
+        incorporate the whole input string, not just one word from it.
+        '''
+        string = self.string2list(string)
+        seed = random.choice(string)
+        return self.next_word(seed)
+
+    def next_word(self, input_word, backward=0):
+        '''
+        1. sum all weights
```
Comment (n0arch) [inline markov.py]: the 1, 2 and 3 should probably just be single line comments where these things actually happen in the function

File: markov.py [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+        r = random.randint(0, s-1)
+
+        for key in self.brain[input_word].iterkeys():
+            if r < self.brain[input_word][key][backward]:
+                return key
+            else:
+                r -= self.brain[input_word][key][backward]
+        return random.choice(self.brain.keys())
+
+    def update_weights(self, string):
```
Comment (n0arch) [inline markov.py]: add docstring

File: markov.py [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+        string = self.string2list(string)
+        if len(string) is 1:
+            return
+        d = [1, -1]
```
Comment (n0arch) [inline markov.py]: maybe add some comments from here down explaining the logic

File: markov.py [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+        new_next = [[1, 0], [0, 1]]
+        for j in [0, 1]:
```
Comment (n0arch) [inline markov.py]: j and i are not v descriptive var names :x

File: markov.py [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+            string = string.rstrip('\n')
+            string = string.decode('utf-8', 'replace')
+            string = string.split(' ')
+            return string
+        elif isinstance(string, list):
+            return string
+
+    def seed_word(self, string):
+        '''
+        algorithm for selecting seed word from given string
```
Comment (n0arch) [inline markov.py]: :+1:

File: markov.py (hunk 2 above)
Comment (n0arch) [inline markov.py]: actually keep them here but add some below where necessary

:+1:

File: markov.py [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+        length_aft = random.randint(1, 10)
+        sentence = [self.seed_word(string)]
+
+        for i in range(length_fwd):
+            word = self.next_word(sentence[-1])
+            if word is None:
+                break
+            sentence.append(word)
+
+        for i in range(length_aft):
```
Comment (n0arch) [inline markov.py]: :+1: length_aft

File: markov.py [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+                    new_current = {string[i+d[j]]: new_next[j]}
+                    if string[i] in self.brain.iterkeys():
+                        if string[i+d[j]] in self.brain[string[i]].iterkeys():
+                            self.brain[string[i]][string[i+d[j]]][j] += 1
+                        else:
+                            self.brain[string[i]][string[i+d[j]]] = new_next[j]
+                    else:
+                        self.brain[string[i]] = new_current
+
+    def generate_sentence(self, string):
```
Comment (n0arch) [inline markov.py]: docstring

File: markov.py [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+
+        for i in sentence:
+            if i is None:
+                sentence.pop(sentence.index(i))
+
+        if sentence is []:
+            return "i'm not sure what to say to that"
+
+        self.write()
+        return ' '.join(sentence)
```
Comment (n0arch) [inline markov.py]: should we add some logic in here to make sentences a little less likely to be run-ons and fragments?

(idk how we'd do this)

Comment (n0arch) [inline markov.py]: we can add that as a to-do for later :+1:

File: pybotX.py [hunk 11]
```diff
@@ -39,9 +39,19 @@
...
 PASSWORD = "<PASSWORD>"
-
+# :'str': Path to markov json file
+BRAIN = "pybot.json"
 # Set our spam variable to None
 LAST_REQUEST = None
+# List of nicknames to ignore
+IGNORED_NICKS = [NICKNAME, 'Francis', 'Meatball']
+
+LISTEN_COUNTER = 0
```
Comment (n0arch) [inline pybotX.py]: wat is this?

File: pybotX.py [hunk 12]
```diff
@@ -39,9 +39,19 @@
...
+# Number of lines the bot will read before it speaks by itself
+TALK_FREQUENCY = 100
+
+m = Markov(BRAIN)
```
Comment (n0arch) [inline pybotX.py]: maybe use more descriptive var name

File: pybotX.py [hunk 13]
```diff
@@ -129,17 +157,29 @@ def _check_spam(request_time):
...
-    except ValueError:
-        repl = markov.generate_markov_text()
-    f.close()
-    return repl
+def markov(ss, mode):
+    global m
+    global LISTEN_COUNTER
+
+    if mode == 'talk':
+        ss = ss.split(' ')
```
Comment (n0arch) [inline pybotX.py]: maybe change `ss` to `seed_string`

**Comment by n0arch** [comment]
some minor nitpicks but looks good 2 me

tumblr_mluhisx54k1s0ybymo1_500

**Comment by n0arch** [comment]
Awesome. 

LGTM

**Pull request merged**